Mutexmap

elena-kolevska opened the pull request:
# Description

This PR adds a map of mutexes that provide an inner and outer locking mechanism to support keeping separate locks on dynamic objects, as is the case with namespaced stream pools, for example.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests

ItalyPaleAle commented:
In general, the code seems good... But what is this going to be used for? We already have other libraries used in the codebase that support thread-safe maps and do that in a more performant way too. For example, we already use [`github.com/alphadose/haxmap`](https://github.com/alphadose/haxmap) in other parts of Dapr ([go.mod here](https://github.com/dapr/kit/blob/ccffb60016c7db4ade97ef758372257d5fa29e32/go.mod#L6)). Although it has _some_ caveat, it's a lot faster, thanks to being lock-free. Could that be used here too?

cicoyle commented on concurrency/atomicmap_test.go:
```diff
@@ -0,0 +1,65 @@
+package concurrency
```
```suggestion
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package concurrency
```

elena-kolevska replied:
Thanks Cassie

cicoyle commented on concurrency/mutexmap_test.go:
```diff
@@ -0,0 +1,91 @@
+package concurrency
```
```suggestion
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package concurrency
```

JoshVanL commented on concurrency/mutexmap.go:
```diff
@@ -0,0 +1,91 @@
+	"sync"
+)
+
+type MutexMap[T comparable] struct {
+	lock  sync.RWMutex
+	items map[T]*sync.RWMutex
+}
```
Make this a non-exported struct type, and create an interface implementing the below funcs. This prevents consumers creating base `MutexMap` structs which will not have `items` instantiated.

JoshVanL commented on concurrency/mutexmap.go:
```diff
@@ -0,0 +1,91 @@
+	lock  sync.RWMutex
+	items map[T]*sync.RWMutex
+}
+
```
func NewMutexMap[T comparable]() *MutexMap[K] {
	return &MutexMap[string]{
		items: make(map[K]*sync.RWMutex),
	}
}

JoshVanL commented on concurrency/atomicmap.go:
```diff
@@ -0,0 +1,121 @@
+		items: make(map[string]*AtomicValue[uint32]),
+	}
+}
+
```
```suggestion
func NewAtomicMap[T comparable, K constraints.Integer]() *AtomicMap[T, K] {
	return &AtomicMap[T, K]{
		items: make(map[T]*AtomicValue[K]),
	}
}
```

elena-kolevska commented:
Thanks @ItalyPaleAle , haxmap looks great, I wasn't aware of it. I updated the PR in runtime to use it instead and removed the atomic map from kit. I'd like to keep the mutexmap though, because it has some more specialised methods that keep the code in runtime cleaner.

JoshVanL commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+**/.DS_Store
+.idea
+.vscode
+.vs
```
nit: newline

JoshVanL commented on concurrency/mutexmap.go:
```diff
@@ -0,0 +1,107 @@
+import (
+	"sync"
+)
+
```
Nit: add comment describing interface

elena-kolevska replied:
Done! Thanks.

JoshVanL commented on concurrency/mutexmap.go:
```diff
@@ -0,0 +1,120 @@
+// - Delete(key T): Removes the mutex associated with the given key from the map.
+// - Clear(): Removes all mutexes from the map.
+// - ItemCount() int: Returns the number of items (mutexes) in the map.
+
```
nit nit :slightly_smiling_face: 

```suggestion
```

JoshVanL replied:
(this may bork some IDEs/doc generators).

elena-kolevska replied:
Ah yes, sorry, Goland does a bunch of formatting, and sometimes it's not easy to recognise it's an actual empty line, and not its own formatting.

ItalyPaleAle commented:
> Thanks @ItalyPaleAle , haxmap looks great, I wasn't aware of it. I updated the PR in runtime to use it instead and removed the atomic map from kit

Just keep in mind one important thing about haxmap. If an item is added to the map _while it's being resized internally_, the item shows up only after the resizing is done. [See here](https://github.com/alphadose/haxmap/issues/32#issuecomment-1730752209)

Would this be a problem?

elena-kolevska commented:
> If an item is added to the map while it's being resized internally, the item shows up only after the resizing is done. 

Thanks for pointing that out. Does this also relate to GetOrSet? Will it block until the resize is complete, or?
Also... the issue doesn't seem to have reached a consensus on the cause of this error. Someone else suggested it's related to deletes. Do you remember more details? I know it's been a while, but I think this is important to get right 🙏

ItalyPaleAle commented:
It won't block. How haxmap works is that it allocates memory to store the items. If more items are needed than the memory that's available, it will need to allocate a new block of memory, copy the items, and then move the pointer to the new memory.

Haxmap doesn't block. As the docs say:

> If a resizing operation is happening concurrently while calling Set() then the item might show up in the map only after the resize operation is finished

If you know in advance how many items you'll have, you can pre-allocate the haxmap with enough capacity for them.

elena-kolevska commented:
Unfortunately, we don't know the number of elements, it's going to depend on how users use the system (specifically, for now, it will be the number of namespaces). I suppose we could set the initial map size to a higher starting number of ~50, or  make the size a placement service start up parameter, but I don't really like either of these options.

I'm concerned that a non-replicated data structure can't give me a read-your-write consistency. I didn't dig deep into the paper and the inner structure, and I'm sure if it were easy, it would have been done, but this doesn't look like an unsolvable problem. For example, the resize could be triggered in the background when the size reaches X% of full capacity, and even if we still get to a point where a write has to happen before the resize completes, the write operation shouldn't be acknowledged to the user before the resize has completed, for correctness.

Since we can have multiple sidecars from the same namespace trying to send their updates at the same time, we might indeed have a situation where this issue affects us.

Let me know if I'm understanding this correctly, or missing something.

elena-kolevska commented:
Looking at [this issue](https://github.com/alphadose/haxmap/issues/26) now, there also seems to be a race on the `GetOrSet` operation, where a delete wouldn't be observed if it's concurrent to a GetOrSet. It's still eventually consistent, but we need strong consistency here. 
If everyone's ok with it, I would like to move back to the sync.Mutex based map we had before, even though it's sad to give up on the performance benefits of haxmap.
The library is really interesting though, and maybe we could contribute to it in the future and address these issues.

artursouza commented:
+1 to use sync.map.

elena-kolevska commented:
Thanks @artursouza , I reverted it.

ItalyPaleAle commented on concurrency/atomicmap.go:
```diff
@@ -0,0 +1,111 @@
+func (a *atomicMap[K, T]) Clear() {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	a.items = make(map[K]*AtomicValue[T])
```
Use clear here please

ItalyPaleAle commented on concurrency/atomicmap.go:
```diff
@@ -0,0 +1,111 @@
+	"golang.org/x/exp/constraints"
+)
+
+type AtomicValue[T constraints.Integer] struct {
```
For this struct, why not using `sync.Atomic` (like `atomic.Int32`) instead? That's lock-free and should have a lot better performance.

elena-kolevska replied:
That was the initial idea, but one of the way I'm using it now is with Int64, and another one as UInt32, so it needs to be flexible. We looked at using `atomic.Value`, but it doesn't implement `Add()`, which we need.

ItalyPaleAle replied:
I'd recommend creating 2 structs then, the use of locking is orders of magnitude slower than using atomic ([some source](https://stackoverflow.com/a/47446114/192024))

There's definitely an Add: https://pkg.go.dev/sync/atomic#Int32.Add

elena-kolevska replied:
That's a good point! I was curious and did a little benchmark of an atomic vs mutex add:

```
BenchmarkAtomicAdd-12    	21512379	        59.52 ns/op	       0 B/op	       0 allocs/op
BenchmarkMutexAdd-12     	10634985	       113.2 ns/op	       0 B/op	       0 allocs/op
```

So maybe not orders of magnitude, but the mutex-based one is almost twice as slow.

So I guess it's a question of "dirtying up" the code and getting the performance benefit, or keeping it cleaner/more maintainable, but paying the latency penalty.
I don't have a strong opinion here, to be honest, but I usually lean more towards maintainability, unless it's a really high-load system with operations per second in the hundreds of thousands. What do you all think?

artursouza replied:
I spent some time exploring the alternative to workaround the generics to handle both cases with a single implementation but I got blocked by limited typecast with generics in Go (I cannot do `return (T) myInt64`, for example)

I do agree that duplicating the map seems overkill unless there is a justified use case that demands it - classic premature optimization case for now IMO.

This is me adding the Int64 variant of this map (for future reference)
https://github.com/artursouza/kit/commit/94fa3f4d82c1882f5801abcb4be26dcac7b4d4ae

Unless there are more data points, I recommend we merge this code with a single implementation of the atomic map.

JoshVanL replied:
I agree, I think this is ok to merge as is. I think the usage of this package will not likely be a limiting factor for performance and the benchmark diff isn't worth the code sprawl.

ItalyPaleAle commented:
@artursouza you merged this PR when there was only 1 vote. The comment should not have been resolved as the conversation was not over.

If the problem was the need to support both int64 and uint32, you don't need to duplicate the code. You can still use the atomic pattern, internally using only `atomic.Int64`. Since uint32 can safely be contained within int64, you can still use a generic class and then cast the response appropriatley.

artursouza commented:
> @artursouza you merged this PR when there was only 1 vote. The comment should not have been resolved as the conversation was not over.
> 
> If the problem was the need to support both int64 and uint32, you don't need to duplicate the code. You can still use the atomic pattern, internally using only `atomic.Int64`. Since uint32 can safely be contained within int64, you can still use a generic class and then cast the response appropriatley.

Can you provide the code for that? I did hit limitations in the casting capabilities of Go - maybe there is a viable path. I am happy to take another PR on this but I don't think this possible improvement justifies blocking progress on 1.14, given the deadline.
